dynamic-check/src/assertions/type_variable_pair.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::TypeAssertionError;
+use leo_static_check::{flatten_array_type, Type, TypeVariable};
+use leo_typed::Span;
+
+/// A type variable -> type pair.
+pub struct TypeVariablePair(TypeVariable, Type);
+
+impl TypeVariablePair {
+    pub fn first(&self) -> &TypeVariable {
+        &self.0
+    }
+
+    pub fn second(&self) -> &Type {
+        &self.1
+    }
+}
+
+/// A vector of `TypeVariablePair`s.
+pub struct TypeVariablePairs(Vec<TypeVariablePair>);
+
+impl Default for TypeVariablePairs {
+    fn default() -> Self {
+        Self(Vec::new())
+    }
+}
+
+impl TypeVariablePairs {
+    ///
+    /// Returns a new `TypeVariablePairs` struct from the given left and right types.
+    ///
+    pub fn new(left: &Type, right: &Type, span: &Span) -> Result<Self, TypeAssertionError> {
+        let mut pairs = Self::default();
+
+        // Push all `TypeVariablePair`s.
+        pairs.push_pairs(left, right, span)?;
+
+        Ok(pairs)
+    }
+
+    ///
+    /// Returns true if the self vector has no pairs.
+    ///
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    ///
+    /// Returns the self vector of pairs.
+    ///
+    pub fn get_pairs(&self) -> &Vec<TypeVariablePair> {
+        &self.0
+    }
+
+    ///
+    /// Pushes a new `TypeVariablePair` struct to self.
+    ///
+    pub fn push(&mut self, variable: &TypeVariable, type_: &Type) {
```

Comment:
Rust APIs usually pass `push` arguments by value, leaving the cloning to the caller